cmds/info.c

```diff
@@ -853,6 +853,116 @@ static int read_uftrace_version(void *arg)
 	return 0;
 }
 
+void write_header(int fd, struct uftrace_file_header *hdr)
+{
+	dprintf(fd, "magic:%s\n", hdr->magic);
+	dprintf(fd, "version:%u\n", hdr->version);
+	dprintf(fd, "header_size:%u\n", hdr->header_size);
+	dprintf(fd, "endian:%u\n", hdr->endian);
+	dprintf(fd, "elf_class:%u\n", hdr->elf_class);
+	dprintf(fd, "feat_mask:%lu\n", hdr->feat_mask);
+	dprintf(fd, "info_mask:%lu\n", hdr->info_mask);
+	dprintf(fd, "max_stack:%u\n", hdr->max_stack);
+}
+
+static int read_magic(struct uftrace_file_header *hdr, FILE *fp)
+{
+	char buf[PATH_MAX];
+	if (fgets(buf, PATH_MAX, fp) == NULL)
+		return -1;
+	if (strncmp(buf, "magic:", 6))
+		return -1;
+	memcpy(hdr->magic, copy_info_str(&buf[6]), UFTRACE_MAGIC_LEN - 1);
+	return 0;
+}
+
+static int read_version(struct uftrace_file_header *hdr, FILE *fp)
+{
+	char buf[PATH_MAX];
+	if (fgets(buf, PATH_MAX, fp) == NULL)
+		return -1;
+	if (strncmp(buf, "version:", 8))
+		return -1;
+	sscanf(&buf[8], "%u", &hdr->version);
+	return 0;
+}
+
+static int read_header_size(struct uftrace_file_header *hdr, FILE *fp)
+{
+	char buf[PATH_MAX];
+	if (fgets(buf, PATH_MAX, fp) == NULL)
+		return -1;
+	if (strncmp(buf, "header_size:", 12))
+		return -1;
+	sscanf(&buf[12], "%hu", &hdr->header_size);
+	return 0;
+}
+
+static int read_endian(struct uftrace_file_header *hdr, FILE *fp)
+{
+	char buf[PATH_MAX];
+	if (fgets(buf, sizeof(buf), fp) == NULL)
+		return -1;
+	if (strncmp(buf, "endian:", 7))
+		return -1;
+	sscanf(&buf[7], "%hhu", &hdr->endian);
+	return 0;
+}
+
+static int read_elf_class(struct uftrace_file_header *hdr, FILE *fp)
+{
+	char buf[PATH_MAX];
+	if (fgets(buf, sizeof(buf), fp) == NULL)
+		return -1;
+	if (strncmp(buf, "elf_class:", 10))
+		return -1;
+	sscanf(&buf[10], "%hhu", &hdr->elf_class);
+	return 0;
+}
+
+static int read_feat_mask(struct uftrace_file_header *hdr, FILE *fp)
+{
+	char buf[PATH_MAX];
+	if (fgets(buf, sizeof(buf), fp) == NULL)
+		return -1;
+	if (strncmp(buf, "feat_mask:", 10))
+		return -1;
+	sscanf(&buf[10], "%lu", &hdr->feat_mask);
+	return 0;
+}
+
+static int read_info_mask(struct uftrace_file_header *hdr, FILE *fp)
+{
+	char buf[PATH_MAX];
+	if (fgets(buf, sizeof(buf), fp) == NULL)
+		return -1;
+	if (strncmp(buf, "info_mask:", 10))
+		return -1;
+	sscanf(&buf[10], "%lu", &hdr->info_mask);
+	return 0;
+}
+
+static int read_max_stack(struct uftrace_file_header *hdr, FILE *fp)
+{
+	char buf[PATH_MAX];
+	if (fgets(buf, sizeof(buf), fp) == NULL)
+		return -1;
+	if (strncmp(buf, "max_stack:", 10))
+		return -1;
+	sscanf(&buf[10], "%hu", &hdr->max_stack);
+	return 0;
+}
+
+int read_header(struct uftrace_file_header *hdr, FILE *fp)
+{
+	if (read_magic(hdr, fp) < 0 || read_version(hdr, fp) < 0
+		|| read_header_size(hdr, fp) < 0 || read_endian(hdr, fp) < 0
+		|| read_elf_class(hdr, fp) < 0 || read_feat_mask(hdr, fp) < 0
+		|| read_info_mask(hdr, fp) < 0 || read_max_stack(hdr, fp) < 0)
+		return -1;
+	return 0;
+}
```

Comment:
There are too many repetition here. I think we can use a single common function and this can be shorter as follows.
```c
static char *read_header_field(struct uftrace_file_header *hdr, FILE *fp,
                   const char *field, char *buf)
{
    int len = strlen(field);

    if (fgets(buf, PATH_MAX, fp) == NULL)
        return NULL;
    if (strncmp(buf, field, len))
        return NULL;
    return &buf[len];
}

int read_header(struct uftrace_file_header *hdr, FILE *fp)
{
    char buf[PATH_MAX];
    char *data;

    data = read_header_field(fp, hdr, "magic:", buf);
    if (!data)
        return -1;
    memcpy(hdr->magic, copy_info_str(data), UFTRACE_MAGIC_LEN - 1);

    data = read_header_field(fp, hdr, "version:", buf);
    if (!data)
        return -1;
    sscanf(data, "%u", &hdr->version);

    data = read_header_field(fp, hdr, "header_size:", buf);
    if (!data)
        return -1;
    sscanf(data, "%hu", &hdr->header_size);

    data = read_header_field(fp, hdr, "endian:", buf);
    if (!data)
        return -1;
    sscanf(data, "%hhu", &hdr->endian);

    data = read_header_field(fp, hdr, "elf_class:", buf);
    if (!data)
        return -1;
    sscanf(data, "%hhu", &hdr->elf_class);

    data = read_header_field(fp, hdr, "feat_mask:", buf);
    if (!data)
        return -1;
    sscanf(data, "%"PRIu64, &hdr->feat_mask);

    data = read_header_field(fp, hdr, "info_mask:", buf);
    if (!data)
        return -1;
    sscanf(data, "%"PRIu64, &hdr->info_mask);

    data = read_header_field(fp, hdr, "max_stack:", buf);
    if (!data)
        return -1;
    sscanf(data, "%hu", &hdr->max_stack);

    return 0;
}
```